lsp/src/uri0.ml

```diff
@@ -1,31 +1,137 @@
+(* This module is based on the [vscode-uri] implementation:
+   https://github.com/microsoft/vscode-uri/blob/main/src/uri.ts. It only
+   supports scheme, authority and path. Query, port and fragment are not
+   implemented *)
+
 open Import
 
 module Private = struct
   let win32 = ref Sys.win32
 end
 
-type t = Uri_lexer.t =
-  { scheme : string option
+type t =
+  { scheme : string
   ; authority : string
   ; path : string
   }
 
-let t_of_yojson json = Json.Conv.string_of_yojson json |> Uri_lexer.of_string
+let backslash_to_slash =
+  String.map ~f:(function
+    | '\\' -> '/'
+    | _ as c -> c)
+
+let slash_to_backslash =
+  String.map ~f:(function
+    | '/' -> '\\'
+    | _ as c -> c)
+
+let of_path path =
+  let path = if !Private.win32 then backslash_to_slash path else path in
+  let path, authority =
+    let len = String.length path in
+    if len = 0 then ("/", "")
+    else if len > 1 && path.[0] = '/' && path.[1] = '/' then
+      let idx = String.index_from_opt path 2 '/' in
+      match idx with
+      | None -> ("/", String.sub path ~pos:2 ~len:(len - 2))
+      | Some i ->
+        let authority = String.sub path ~pos:2 ~len:(i - 2) in
+        let path =
+          let path = String.sub path ~pos:i ~len:(len - i) in
+          if String.is_empty path then "/" else path
+        in
+        (path, authority)
+    else (path, "")
+  in
+  let path = if path.[0] <> '/' then "/" ^ path else path in
+  { scheme = "file"; authority; path }
+
+let to_path { path; authority; scheme } =
+  let path =
+    let len = String.length path in
+    if len = 0 then "/"
+    else if (not (String.is_empty authority)) && len > 1 && scheme = "file" then
+      "//" ^ authority ^ path
+    else if len < 3 then path
+    else
+      let c0 = path.[0] in
+      let c1 = path.[1] in
+      let c2 = path.[2] in
+      if
+        c0 = '/'
+        && ((c1 >= 'A' && c1 <= 'Z') || (c1 >= 'a' && c1 <= 'z'))
+        && c2 = ':'
+      then
+        String.make 1 (Char.lowercase_ascii c1)
+        ^ String.sub path ~pos:2 ~len:(String.length path - 2)
+      else path
+  in
+  if !Private.win32 then slash_to_backslash path else path
+
+let of_string s =
+  let re =
+    Re.Perl.re "^(([^:/?#]+?):)?(\\/\\/([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?"
```

Comment:
I copied this regular expression from the upstream implementation. If you prefer, I can try to create a parse with menhir

Reply:
Why do we need menhir? It seems like it should be easy enough to convert this regex to ocamllex.

Reply:
Oh yes that's what I meant sorry I got confused

Reply:
The performance should be essentially the same. It's just that with ocamllex we don't need to read line noise and the capturing groups are named rather than numbered.